---
title: "heatmaps vignette"
author: "Malcolm Perry"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

# heatmaps tutorial

## Flexible plotting of Functional Genomics data + Sequence features

This package provides simple functions for plotting heatmaps over sets 
of genomic windows. It provides simple functions for common operations while
still allowing the user full control over plotting options.

## Heatmap class

The Heatmap class contains all the information required to plot a heatmap. It
doesn't contain information about the data used to create the information, or
any options concerning how the heatmap will be plotted.

```{r}
library(heatmaps)

hm = Heatmap(
    image=matrix(0, nrow=2000, ncol=400), 
    coords=c(-200, 200), 
    scale=c(0, 1),
    nseq=2000, 
    label="Example")

```

The most important part of a heatmap is the `image`, a numeric matrix. Note that
the size of `image` is not constrained - it is often a good idea to condense
information for wide windows or large numbers of sequences (how to do this will
be shown later). 

The coords field is optional, and specifies the co-ordinates for our heatmap. More
usually, heatmaps will be centered on some feature rather than starting at 0. In this
example, our region is Â±200bp around our features of interest. Co-ords usually overlap
zero, but they don't have to.

The scale keeps track of the range of data to display on the heatmap. It will
usually be equivalent to `c(0, max(image(hm)))`, but in cases with negative
values it can change, and it can be used to avoid plotting extreme values, in
ways which will be explained later.

The nseq feature keeps track of the number of input sequences, if desired. The label
is passed to the plotting function which can optionally plot text labels onto heatmaps.

Calling `str()` reveals one more field contained in Heatmaps, `metadata`. This is often
included in Bioconductor classes in case users want to keep track of additional data
in classes. Here it could be used to track sample metadata, for example, which could
be passed to plotting functions which might alter the color scheme based on the data.
By default it is empty:

```{r}
str(hm)
```

## Basic Plotting

First we need some data to plot. In this example, we will plot the density of various
normal distributions centered on zero with a standard deviation from 5 to 50. Of course,
most of the time you will be plotting real biological data, but if you are not using
data for which useful helper functions exist (which will be covered later!) then you'll
have to manually create image matrices and feed them to the constructor.

```{r, fig.height=5, fig.width=5}
sd = seq(5, 50, length.out=500)

sample_row = function(sigma) {
    dnorm(-100:100, 0, sd=sigma)
}

img = t(sapply(sd, sample_row))

hm = Heatmap(img, coords=c(-200, 200), label="ND")

plotHeatmap(hm)
box("figure") # to visualize figure margins

```

You'll notice that the aspect ratio of the plot is a bit weird and the margins aren't
even in the plot above. `plotHeatmap` does not actually control the device parameters,
it just plots to the current device. This gives us greater control over how our plots
turn out and means we can easily handle things like multiple plots in one image using
base R (rather than hard-coded options).

```{r, fig.height=5, fig.width=5}

layout(matrix(1:2, nrow=1))
par(mai=c(0.6, 0.3, 0.1, 0.3))
plotHeatmap(hm)
box("figure")
plotHeatmap(hm, color="Reds")
box("figure")

```

The example above also shows the argument `color` which is passed to plotHeatmap.
The default schemes are implemented through RColorBrewer: a full list is available
at http://colorbrewer.org. Other colorschemes can be specified as vectors of colors
which are passed to `colorRampPalette` to generate a smooth colorscheme.

There are many more plotting options available: more will be highlighted in this
vignette, and for a full list see ?heatmapOptions.

## Coverage Heatmaps

Many genome-wide NGS-based experiments generate results which can be visualised as
coverage tracks along the genome, such as wiggle (.wig) or bedGraph files. We can
visualise this signal across a set of known features in a coverage heatmap.

First we need to load some data. We will use a set of zebrafish promoters
as defined by CAGE (Haberle et al., 2014, Nature), and a set of CpG islands
for zebrafish.

```{r}
library(GenomicRanges)
cage = read.table(system.file("extdata", "zf_proms.txt", package="heatmaps"), header=TRUE)
cgi = read.table(system.file("extdata", "zf_cgi.txt", package="heatmaps"), header=TRUE)
```

We can extract coverage from the CpG island object, and make our promoters into windows.
We need the correct seqinfo otherwise `coverage` does not know how long chromosomes should
be, and will throw an error if we try and read from after the end.

```{r}
library(BSgenome.Drerio.UCSC.danRer7)
cage_gr = makeGRangesFromDataFrame(cage, keep.extra.columns=TRUE, seqinfo=seqinfo(Drerio))
cage_gr = cage_gr[order(cage_gr$interquantileWidth)]
windows = promoters(cage_gr, 500, 500)

cgi_gr = makeGRangesFromDataFrame(cgi, keep.extra.columns=TRUE, seqinfo=seqinfo(Drerio))
cov_rle = coverage(cgi_gr, weight="obsExp")
```

And using these two objects, we can create and plot our coverage heatmap:

```{r, fig.height=5, fig.width=2.5}
coverage_hm = CoverageHeatmap(windows, cov_rle, coords=c(-500, 500), label="CGI")
scale(coverage_hm) = c(0, 2)

par(mai=c(0.6, 0.3, 0.1, 0.3))
plotHeatmap(coverage_hm, color="YlOrRd")
```

## Pattern Heatmaps

Another way to investigate CG content in Zebrafish promoters would be to directly visualise
CpG content. heatmaps contains tools to easily extract sequence features from genome sequence,
which is an interesting addition to many analyses.

```{r, fig.height=5, fig.width=2.5}
zf_seq = getSeq(Drerio, windows)
cg_hm = PatternHeatmap(zf_seq, pattern="CG", coords=c(-500, 500), label="CG")

par(mai=c(0.6, 0.3, 0.1, 0.3))
plotHeatmap(cg_hm)
```

Visualising nucleotide patterns is much more effective if we smooth the image
first, rather than drawing the binary matrix.

```{r, fig.height=5, fig.width=2.5}
cg_hm_smooth = smooth(cg_hm)

par(mai=c(0.6, 0.3, 0.1, 0.3))
plotHeatmap(cg_hm_smooth)
```

## Plotting Lists of Heatmaps

Having to set device parameters and plot heatmaps individually is hardly ideal, and
heatmaps contains a function to help with this: `plotHeatmapList()`. Using this we can
easily plot the CG dinucleotide plot alongside the CpG islands to see how they compare:

```{r}
cgi_hm_smooth = smooth(coverage_hm)
scale(cgi_hm_smooth)[1] = 0
plotHeatmapList(list(cg_hm_smooth, cgi_hm_smooth), color=c("Blues", "YlOrRd"))
```

We can see from this analysis that CpG Islands actually do quite a poor job of capturing
CG dinucleotide frequency in Zebrafish: maybe the concept is not as useful as it is in humans,
or our parameters for calling CGIs were incorrect.

